### PR TITLE
GH-2316 Adds d3-tip to map/Pins

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes GH-2316

@GordonSmith please review

### CODE REVIEW

File: src/map/Pins.js [hunk 1]
```diff
@@ -74,17 +75,22 @@
...
+            })
+        ;
     };
 
     Pins.prototype.layerUpdate = function (base) {
         Layer.prototype.layerUpdate.apply(this, arguments);
 
         this._pinsTransform
-            .style("opacity", this.opacity())
+        .style("opacity", this.opacity())
```
Comment (jbrundage) [inline src/map/Pins.js]: @GordonSmith These indentation changes are somehow happening via VSCode. Have you experienced this issue before?

File: src/map/Pins.js [hunk 2]
```diff
@@ -104,6 +110,16 @@
                 }
             })
         ;
+        gPinEnter
+            .each(function(d){
+                if(d.ext && d.ext.tooltip){
```
Comment (GordonSmith) [inline src/map/Pins.js]: While unlikely a data row may start off with no tooltip data, but may get some later on - so you should add the event to every pin regardless.  I may mean we need some more logic elsewhere.

File: src/map/Pins.js [hunk 3]
```diff
@@ -74,17 +75,22 @@
...
     Pins.prototype.layerEnter = function (base, svgElement, domElement) {
         Layer.prototype.layerEnter.apply(this, arguments);
-
+        
         this._pinsTransform = svgElement;
         this._selection.widgetElement(this._pinsTransform);
         this.pinsPaths = d3.select(null);
+        this
+            .tooltipHTML(function (d) {
+                return d.ext.tooltip === undefined ? "" : d.ext.tooltip;
```
Comment (GordonSmith) [inline src/map/Pins.js]: Could show the label if there was no specific tooltip?

File: src/map/Pins.js (hunk 1 above)
Comment (GordonSmith) [reply src/map/Pins.js]: FWIW I turn off "format on save" for non VS Code projects (File/settings -> workspace).  Once you get to 2.0, they have all been reformatted for vscode.

Also you can add the ?w=1 option to the address bar to ignore the whitespace changes.

File: src/map/Pins.js (hunk 3 above)
Comment (jbrundage) [reply src/map/Pins.js]: I don't think the GMapPin data has a label?

File: src/map/Pins.js (hunk 1 above)
Comment (jbrundage) [reply src/map/Pins.js]: Which address bar? The git repo I'm pointing to while pushing?

File: src/map/Pins.js (hunk 2 above)
Comment (jbrundage) [reply src/map/Pins.js]: I believe the pin data points are considered "new" if any aspect of them changes (line ~96 in Pins.js), so "enter" status should apply to the pins after a tooltip is added.

Comment (jbrundage) [reply src/map/Pins.js]: Would this call to **.data()** key the data rows on index? or on the row's content?
```javascript
this.pinsPaths = this._pinsTransform.selectAll(".pin").data(this.visible() ? this.pinsData() : []);
```

File: src/map/Pins.js (hunk 3 above)
Comment (GordonSmith) [reply src/map/Pins.js]: ...true...

File: src/map/Pins.js (hunk 1 above)
Comment (GordonSmith) [reply src/map/Pins.js]: The one pointing to this page ^^^^

File: src/map/Pins.js (hunk 2 above)
Comment (GordonSmith) [reply src/map/Pins.js]: When no "id" function is passed into the ```data``` function on line ~96 it uses the idx position of the data row as its identifier.

**Comment by jbrundage** [comment]
@GordonSmith ready to be rereviewed

**Pull request merged**